### ISSUE TITLE
support either pfx or key/cert/ca combo..

### ISSUE THREAD

**Issue by ohritz** [preciofishbone/OmniaFx-Demos#4]
How about extend the tooling in the task for omnia-serve to handle either pfx(path & password) or the key/cert/ca inputs which webpack-dev-server also supports..

There are npm packages
- https://www.npmjs.com/package/dev-cert-authority
- https://www.npmjs.com/package/selfsigned
which could be used to write a little tooling to generate self signed certs.

With some instructions in the readme, and a gulp task called setup, we could generate certs for the project, or use previously created ones? If the iisexpress.pfx is not present.
same setup task could also ask questions and fill in configurations.

I think it would be easier to build and maintain than a gui wizard in VS..

**Comment by andtii** [comment]
@khdi Did some testing on auto generating certs but chrome was not happy. Can you have a look at  @ohritz suggestions @khdi maybe they will work?

**Comment by ohritz** [comment]
The dev-cert-authority package would require the dev to install i globaly, and it allows creation of a local CA cert which would get installed in the trusted certificate ctore I imagine, which I think would satisfy chrome.

In a previous non Omnia project I setup a  CA server in a dev env on VM's and generated server certs from the CA, and added the CA to my Trusted Root CA's and I could browse to the VM's servers from outside, and get a green padlock in chrome and no issues.
As I recall, it was important that the certs were created with correct common name and alternative name.

**Comment by khdi** [comment]
@ohritz I tried out the package dev-cert-authority but ran into some issues. Seems like that package is not very windows-friendly. When you set up your CA server before did you use Windows or Linux/MacOS?

**Comment by ohritz** [comment]
That was a completly different situation, I set up a sharePoint onPrem dev farm with a dedicated AD machine. All Windows. 
On the Ad machine I set up a certificate authority server, generated a root CA and created a template which I could then request from certificate managers on the addin server and SharePoint server to get a dedicated ssl cert for them.
Then trusting the root ca in my own (host) machine, allowed me to get green padlock ssl in chrome.

The takeaway from that scenario, is having a localhost cert generated from a trusted root cert, with Subject Name { CN = localhost} 
and 
Subject Alternative Name {DNS Name=localhost, URL=localhost}

Gets you through on Chromes new restrictions.

I'm not entirley sure that we need a trusted root cert, but I do know that we need to set subject name and Subject Alternative name, and the cert has to be added to the certificate store.

If we are only supporting windows dev for this. we could write some powershell for that. and call it from a npm script or node process.

I have a powershell script for that.. Will add it to my fork of OmniaFx-Demos
It creates a pfx file and adds it to the cert store.

**Comment by andtii** [comment]
@ohritz Nice! We want Omnia Dev everywhere not just windows :)

**Comment by ohritz** [comment]
I thought so..
But is'nt pfx a windows thing?
In all my dev on linux I've always used the cert and key files.

I'm going to do some testing with the packages mentioned above and maybe some own custom stuff in a pure simple webpack setup and see if I can get something working.

**Comment by andtii** [comment]
@ohritz We will add more cert options that webpack supports, good if you can help solve auto generating certs cross plat

**Comment by ohritz** [comment]
Got a solution working in a pure webpack/vue project..  adding certs working for windows currently.. but generation working via npm packages which should work cross platform.

will do some cleanup and port over into omniaFx Demo..

**Comment by ohritz** [comment]
So I sent a pull request... will se if I can spin up a linux env and implement the trusting there..
But I have no access to a mac, do we know anyone who could help out?

Also, should we add an option to just generate and let the dev handle the trust issue themselves?

Or should we try to pull in the computers own cert or something and create ours from that one? then trusting should'nt be an issue right?

**Comment by andtii** [comment]
@ohritz 
Im using the linux subsystem for win 10, have you tried it? -https://docs.microsoft.com/en-us/windows/wsl/install-win10

If we can automate as much as possible it will be awesome. By the way i havent checked how spfx projects does it but i think they are auto generating the certs.

**Comment by ohritz** [comment]
yeah I got the linux subsytem, not sure if its enough for it to work there in the subsytem, as it does all syscalls to windows.
I've been using ubuntu at home, but recently removed my duelboot.. will try setting up a vm..
There will be requirements on a linux dev to install ca-certificates packet.. f.ex

Also, in Linux there are differences on certificate handling between applications (Browsers).. and not always centraly managed..

**Comment by ohritz** [comment]
will have a look at spFx..

**Comment by ohritz** [comment]
Had a look at spfx, there is a package they use which has gulptasks for generating and trusting certs.. will look into using that instead..
It generates certs and tries to trust them on windows and mac, and instructs the dev to add them to their ca-store on Linux.

**Comment by andtii** [comment]
Nicee that looks exactly what we need!

**Comment by ohritz** [comment]
Jepp.
It looks awsome! :D, works like a charm..
took a bit of digging to figure it out. The build system they have in spFx is several layers deep of abstractions.
But it gives some nice benefits.. 
It might be worth looking into using the core parts of it in the tooling?
They define tasks as classes extending an abstract base class, and then use a build rigg to orchestrate/register and initilize.

**Comment by ohritz** [comment]
but.. It generates pem files and no pfx

**Comment by andtii** [comment]
in that case we could have 
```
DevelopmentEnvironment.hosting
    .useCustom(...)
```

**Comment by ohritz** [comment]
Having it typed is a great Idea.

What I'm suggesting is default being zero config to the build, as we could generate certs, and I see no reason to have to pass in a port. there could be a default internaly.

But if a dev has some "special" case, it shouldnt be to hard to take the default config and
`const config = webpack.merge(defaultConfig, customConfig); `

and its up to the dev to pass in a valid object as per webpacks documentation. And they could tweak it till it works for them..

**Comment by ohritz** [comment]
If we have @types/webpack
and if the dev has thier own cert, they would have to do 
```
DevelopmentEnvironment.hosting
   .use({} :webpackConfig)
```
otherwise
```
DevelopmentEnvironment.hosting
    .useDefault()
```

and the tool makes certs, and spits out on the console what ports and hostname has been used and whatever else info is needed to connect..

**Comment by andtii** [comment]
OK sounds good but one question arrises, since we use the port to connect the extension in the omnia console how does it remember the port between stopping and starting webpack dev server? Currently im having 2,3 extensions running at the same time and want it to be predictable between stopping the server and starting again.

**Comment by ohritz** [comment]
is that really a problem?
if you don't pass anything, and you're running one extension,  It will always use say port, 55433, its the default.. lets say... 
and if you want to run several extenssions..
you run one with default.. and pass in.

```
DevelopmentEnvironment.hosting
   .use({
      devServer: {
           port: 66443 / 77443
      }
})
```
in the other ones..

Its not going to randomly use differnet ports on each start.

**Comment by ohritz** [comment]
I mean the tooling will internally have a set of defaults, not webpacks default necesarily.

**Comment by ohritz** [comment]
mind you, the typings for webpack are not complete.. but the documentation is excellent.

**Comment by andtii** [comment]
Yep sound ok, we will discuss this tomorrow on the review. @khdi any input? @ohritz Will you push an example how we can use the auto cert generating?

**Comment by ohritz** [comment]
I've got an example of the gulp tasks in a branch (https://github.com/ohritz/OmniaFx-Demos/tree/useMsBuildForCerts)

The ms build tool, will generate the certs only as pem.. so I'm doing a bit of jugglery in the gulp task to create a pfx.
Do you want a PR of it? or are yu happy lloking at it in my branch?
